src/puppetlabs/trapperkeeper/bootstrap.clj

```diff
@@ -132,26 +136,102 @@
             (str "Unable to find bootstrap.cfg file via --bootstrap-config "
                  "command line argument, current working directory, or on classpath")))))
 
-(schema/defn chain-files :- [schema/Str]
-  "Takes a list of files, reads all their lines in, and returns a flattened seq
-   of all their lines put together.
+(schema/defn indexed
+  "Returns seq of [index, item] pairs
+  [:a :b :c] -> ([0 :a] [1 :b] [2 :c])"
+  [coll]
+  (map vector (iterate inc 0) coll))
 
-   In the case of an empty file, nil would be returned by line-seq, so we remove
-   all the nils at the end"
-  [files :- [(schema/protocol io/IOFactory)]]
-  (remove nil? (flatten (map #(line-seq (io/reader %)) files))))
+(schema/defn get-annotated-bootstrap-entries :- [AnnotatedBootstrapEntry]
+  [configs :- [(schema/protocol io/IOFactory)]]
+  (flatten (for [config configs]
+             (for [[line-number line-text] (indexed (map remove-comments (line-seq (io/reader config))))
+                   :when (not (empty? line-text))]
+               {:bootstrap-file (str config)
```

Comment:
Agreed that the schema is a little off. I guess the problem I see is that even if I tighten it up to be a string, somewhere else in the code we'll have to assume it has a meaningful `toString` when it gets passed in here. And at some point something will have to be printed in the error messages. I don't think we can make it `File` because if the bootstrap is in a jar file it'll be an `InputStream`.